File: ruby/lib/pagou/logger.rb
# frozen_string_literal: true

require "json"
require "set"

module Pagou
  SENSITIVE_KEYS = %w[
    authorization apikey api_key token access_token client_secret secret
    password cvv cvc pan card_number number
  ].to_set.freeze

  TOKEN_PATTERNS = [
    /\bBearer\s+[A-Za-z0-9._-]+/i,
    /\bpg(ct|pm|sk|pk)_[A-Za-z0-9]+/,
  ].freeze

  REDACTED = "[REDACTED]"

  # Deep-copies a value with sensitive fields masked. Used before anything is
  # logged so secrets, tokens and card data never reach stdout or a log sink.
  def self.redact(value, seen = Set.new)
    case value
    when String
      TOKEN_PATTERNS.reduce(value) { |acc, pattern| acc.gsub(pattern, REDACTED) }
    when Hash
      return "[Circular]" if seen.include?(value.object_id)

      seen.add(value.object_id)
      value.each_with_object({}) do |(key, item), out|
        out[key] = SENSITIVE_KEYS.include?(key.to_s.downcase) ? REDACTED : redact(item, seen)
      end
    when Array
      return "[Circular]" if seen.include?(value.object_id)

      seen.add(value.object_id)
      value.map { |item| redact(item, seen) }
    else
      value
    end
  end

  # A tiny structured logger that redacts before writing. info -> stdout,
  # warn/error -> stderr, matching the reference implementation.
  class Logger
    LEVELS = { info: $stdout, warn: $stderr, error: $stderr }.freeze

    def info(message, context = nil)
      emit(:info, message, context)
    end

    def warn(message, context = nil)
      emit(:warn, message, context)
    end

    def error(message, context = nil)
      emit(:error, message, context)
    end

    private

    def emit(level, message, context)
      line = context ? "#{message} #{JSON.generate(Pagou.redact(context))}" : message
      LEVELS.fetch(level).puts(line)
    end
  end

  def self.logger
    @logger ||= Logger.new
  end
end

File: python/pagou/logger.py
import json
import re
import sys
from typing import Any, Optional

SENSITIVE_KEYS = {
    "authorization",
    "apikey",
    "api_key",
    "token",
    "access_token",
    "client_secret",
    "secret",
    "password",
    "cvv",
    "cvc",
    "pan",
    "card_number",
    "number",
}

_TOKEN_PATTERNS = [
    re.compile(r"\bBearer\s+[A-Za-z0-9._-]+", re.IGNORECASE),
    re.compile(r"\bpg(ct|pm|sk|pk)_[A-Za-z0-9]+"),
]

_REDACTED = "[REDACTED]"


def _redact_string(value: str) -> str:
    out = value
    for pattern in _TOKEN_PATTERNS:
        out = pattern.sub(_REDACTED, out)
    return out


def redact(value: Any, _seen: Optional[set] = None) -> Any:
    """Deep-copies a value with sensitive fields masked.

    Used before anything is logged so secrets, tokens and card data never reach
    stdout or a log sink.
    """
    if isinstance(value, str):
        return _redact_string(value)
    if value is None or not isinstance(value, (dict, list)):
        return value

    if _seen is None:
        _seen = set()
    if id(value) in _seen:
        return "[Circular]"
    _seen.add(id(value))

    if isinstance(value, list):
        return [redact(item, _seen) for item in value]

    out: dict = {}
    for key, item in value.items():
        out[key] = _REDACTED if str(key).lower() in SENSITIVE_KEYS else redact(item, _seen)
    return out


def _emit(stream, message: str, context: Optional[dict]) -> None:
    line = f"{message} {json.dumps(redact(context))}" if context else message
    print(line, file=stream)


class Logger:
    def info(self, message: str, context: Optional[dict] = None) -> None:
        _emit(sys.stdout, message, context)

    def warn(self, message: str, context: Optional[dict] = None) -> None:
        _emit(sys.stderr, message, context)

    def error(self, message: str, context: Optional[dict] = None) -> None:
        _emit(sys.stderr, message, context)


logger = Logger()
